layers/tls_test.go

```diff
@@ -0,0 +1,318 @@
+	0x13, 0xd3, 0x77, 0xc7, 0xc5, 0x79, 0x16, 0x56, 0x4c, 0x30, 0x94, 0xcf, 0xb0, 0x3c, 0x00, 0x91,
+	0xbd, 0x86, 0x08, 0x9f, 0x2f, 0x05, 0x67, 0x03, 0x6f, 0xa7, 0x3b, 0xb9, 0x36, 0xf2, 0x80, 0x4f,
+	0x60, 0x5d, 0x4c, 0xc4, 0x42, 0x5d, 0x02, 0x44, 0xba, 0x31, 0x8f, 0x39, 0x8e, 0x0c, 0x1e, 0xa8,
+	0x26, 0x4f, 0x3e, 0x01, 0x96, 0xb3, 0x6f, 0xc6, 0x25, 0xe4, 0x30, 0x03, 0xd6, 0x3a, 0x7d, 0x14,
+	0x03, 0x01, 0x00, 0x01, 0x01, 0x16, 0x03, 0x01, 0x00, 0x30, 0x25, 0xb8, 0x58, 0xc1, 0xa6, 0x3f,
+	0xf8, 0xbd, 0xe6, 0xae, 0xbd, 0x98, 0xd4, 0x75, 0xa5, 0x45, 0x1b, 0xd8, 0x6a, 0x70, 0x79, 0x86,
+	0x29, 0x4e, 0x4f, 0x64, 0xba, 0xe7, 0x1f, 0xca, 0x4b, 0x96, 0x9b, 0xf7, 0x0b, 0x50, 0xf5, 0x4f,
+	0xfd, 0xda, 0xda, 0xcd, 0xcd, 0x4b, 0x12, 0x2e, 0xdf, 0xd5,
+}
+
+// Packet 13 - Two Application Data Records
+var testDoubleAppData = []byte{
+	0x17, 0x03, 0x01, 0x00, 0x20, 0x77, 0x3a, 0x94, 0x7d, 0xb4, 0x47, 0x4a, 0x1d, 0xd4, 0x6c, 0x5a,
+	0x69, 0x74, 0x03, 0x93, 0x32, 0xca, 0x54, 0x5e, 0xa5, 0x81, 0x99, 0x6a, 0x73, 0x66, 0xbf, 0x06,
+	0xa0, 0xdc, 0x6a, 0x9c, 0xb1, 0x17, 0x03, 0x01, 0x00, 0x20, 0x44, 0x64, 0xc8, 0xc2, 0x5a, 0xfc,
+	0x4a, 0x82, 0xdd, 0x53, 0x6d, 0x30, 0x82, 0x4d, 0x35, 0x22, 0xf1, 0x5f, 0x3b, 0x96, 0x66, 0x79,
+	0x61, 0x9f, 0x51, 0x93, 0x1b, 0xbf, 0x53, 0x3b, 0xf8, 0x26,
+}
+var testDoubleAppDataDecoded = &TLS{
+	BaseLayer: BaseLayer{
+		Contents: testDoubleAppData[37:],
+		Payload:  nil,
+	},
+	ChangeCipherSpec: nil,
+	Handshake:        nil,
+	Appdata: []TLSAppDataRecord{
+		{
+			TLSRecordHeader{
+				ContentType: 23,
+				Version:     0x0301,
+				Length:      32,
+			},
+			testDoubleAppData[5 : 5+32],
+		},
+		{
+			TLSRecordHeader{
+				ContentType: 23,
+				Version:     0x0301,
+				Length:      32,
+			},
+			testDoubleAppData[42 : 42+32],
+		},
+	},
+	Alert: nil,
+}
+
+var testAlertEncrypted = []byte{
+	0x15, 0x03, 0x03, 0x00, 0x20, 0x44, 0xb9, 0x9c, 0x2c, 0x6e, 0xab, 0xa3, 0xdf, 0xb1, 0x77, 0x04,
+	0xa2, 0xa4, 0x3a, 0x9a, 0x08, 0x1d, 0xe6, 0x51, 0xac, 0xa0, 0x5f, 0xab, 0x74, 0xa7, 0x96, 0x24,
+	0xfe, 0x62, 0xfe, 0xe8, 0x5e,
+}
+var testAlertEncryptedDecoded = &TLS{
+	BaseLayer: BaseLayer{
+		Contents: testAlertEncrypted,
+		Payload:  nil,
+	},
+	ChangeCipherSpec: nil,
+	Handshake:        nil,
+	Appdata:          nil,
+	Alert: []TLSAlertRecord{
+		{
+			TLSRecordHeader{
+				ContentType: 21,
+				Version:     0x0303,
+				Length:      32,
+			},
+			0xFF,
+			0xFF,
+			testAlertEncrypted[5:],
+		},
+	},
+}
+
+// Malformed TLS records
+var testMalformed = []byte{
+	0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f,
+	0x51, 0x52, 0x53, 0x54, 0x55, 0x56, 0x57, 0x58, 0x59, 0x5a, 0x5b, 0x5c, 0x5d, 0x5e, 0x5f,
+	0xd1, 0xd2, 0xd3, 0xd4, 0xd5, 0xd6, 0xd7, 0xd8, 0xd9, 0xda, 0xdb, 0xdc, 0xdd, 0xde, 0xdf,
+}
+
+var testTLSDecodeOptions = gopacket.DecodeOptions{
+	SkipDecodeRecovery:       true,
+	DecodeStreamsAsDatagrams: true,
+}
+
+func TestParseTLSClientHello(t *testing.T) {
```

Comment:
You can add a new small test here if you want to make sure the right behavior for the previously described `DecodeFromBytes` is followed:

```go
func TestTLSClientHelloDecodeFromBytes(t *testing.T) {
  var tls TLS
  if err := tls.DecodeFromBytes(tlsClientHello, ...); err != nil {
    t.Fatalf("failed first decode: %v", err)
  }
  if !reflect.DeepEquals(tls, clientHelloDecoded) {
    t.Fatalf("first decode doesn't match")
  }
  if err := tls.DecodeFromBytes(tlsClientHello, ...); err != nil {
    t.Fatalf("failed second decode: %v", err)
  }
  if !reflect.DeepEquals(tls, clientHelloDecoded) {
    t.Fatalf("second decode doesn't match")
  }
}
```